### ISSUE TITLE
Resources created by count iteration are needlessly recreated when a value is inserted into the list

### ISSUE THREAD

**Issue by jaydub** [terraform-providers/terraform-provider-openstack#245]
### Terraform Version
Terraform 0.11.3, Openstack provider 1.2.0

### Affected Resource(s)
openstack_networking_secgroup_rule_v2, but probably any immutable resource. This may really be a terraform core level issue.

### Terraform Configuration Files
```hcl
resource "openstack_networking_secgroup_v2" "rule_test_rules_secgroup" {
  region = "${var.region}"
  name = "rule_test_rules_secgroup"
  description = "Testing terraform 0.11.3/openstack 1.2.0 list handling behaviour"
  delete_default_rules = true
}

locals {
  bunch_of_ips = [
    "192.168.0.5/32",
    "192.168.0.4/32",
    "192.168.0.3/32",
    "192.168.0.2/32",
    "192.168.0.1/32",
    ### New IP added in the middle of the list                                                                                  
    #"10.0.0.1/32",
    ### end                                                                                                                     
    "192.168.0.105/32",
    "192.168.0.104/32",
    "192.168.0.103/32",
    "192.168.0.102/32",
    "192.168.0.101/32",
  ]
  rules_secgroup = "${openstack_networking_secgroup_v2.rule_test_rules_secgroup.id}"
}

# A rule to test with                                                                                                           
resource "openstack_networking_secgroup_rule_v2" "ssh_in" {
  count = "${length(local.bunch_of_ips)}"
  region = "${var.region}"
  direction = "ingress"
  ethertype = "IPv4"
  protocol = "tcp"
  port_range_min = 22
  port_range_max = 22
  remote_ip_prefix = "${element(local.bunch_of_ips, count.index)}"
  security_group_id = "${local.rules_secgroup}"
}
```

### Debug Output
First plan and apply: 
https://gist.github.com/jaydub/db2032c77921b637d544211e3612ce9f

Second apply:
https://gist.github.com/jaydub/ada31b6b0ce5cde5ab6d3a8880570ff6

### Expected Behavior
In an ideal world, terraform would regard the count order as state house keeping, and only create, change or destroy resources as necessary, eg you could reverse the list and nothing would happen, and in this case only one rule would be added.

### Actual Behavior

Two things when wrong here: 

1.  the plan required destroying and recreating all resources after the insertion point. This is probably a terraform core problem
2.  an error in state tracking means that it attempts to create a resource before it's old one is destroyed, and this results in an error that is corrected on a second apply.

Needless to say this is not desirable for security rules, as rules are momentarily missing, and in one case are missing for as long as it takes to run a second apply. 

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. Create the base configuration with the "extra" IP commented out
2. Uncomment the extra IP
3. plan and apply with the extra IP
4. plan and apply a second time to correct the error

**Comment by jtopjian** [comment]
@jaydub Indeed, I would say this is more of a core question than something specific to the OpenStack provider. I would suggest opening an issue at the Terraform repo or perhaps posting a question to the Terraform mailing list - I'm sure there are others who have run into this use-case.

One way around this is to only append to the list, but I can see that might not be possible to always do, especially if the list is generated from outside of Terraform.

> Needless to say this is not desirable for security rules, as rules are momentarily missing, and in one case are missing for as long as it takes to run a second apply.

Unfortunately OpenStack will trigger an error if a duplicate rule is attempted to be created. If it didn't, I would have suggested the `create_before_destroy` lifecycle attribute.

**Comment by jtopjian** [comment]
Revisiting this, this is definitely a core Terraform issue that should be resolved in v0.12 (possibly with the introduction of for loops). I'm going to close this issue since it's not specific to this provider. If you'd like to discuss this further, I recommend opening an issue in github.com/hashicorp/terraform.

Please let me know if you have any questions, though.
